### ISSUE TITLE
documentation LLCP example doesn't work

### ISSUE THREAD

**Issue by nehpetsde** [nfcpy/nfcpy#16]
Hi Stephen

This probably is not an actual bug of the lib itself, but rather a docs problem, yet it heavily affects library usability. Or I'm just doing something wrong. Anyway, I've gone to incorporating NFC into my application, and found that LLCP example not really working. The very last piece of code from here http://nfcpy.readthedocs.org/en/latest/topics/get-started.html#work-with-a-peer says:

    import nfc, nfc.snep, threading
    clf = nfc.ContactlessFrontend('usb')
    connected = lambda llc: threading.Thread(target=llc.run()).start()
    llc = clf.connect(llcp={'on-connect': connected})
    snep = nfc.snep.SnepClient(llc)
    snep.put(nfc.ndef.Message(nfc.ndef.SmartPosterRecord("http://nfcpy.org")))
    clf.close()

I've copied it as is into a separate helloworld.py file expecting it to work as is, but it doesn't. It starts NFC, and connection is recognized by the phone, but URL is not transferred. 

I've tried to look into beam.py for other sources of simple enough example to start with, yet that script seems to be much more complex to start with unfortunately.

So what's wrong with this script from the manual, or what I'm doing wrong?

Regards
Alex

PS: the actual script on that manual page has a typo, a closing quote is missing on the second line around 'usb' string, and this typo is copypasted in few other places on the same manual page.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2013-11-11T23:47:25Z
 * owner: alexykot
 * assignee: stephen-tiedemann
 * the launchpad url was https://bugs.launchpad.net/bugs/1250255

**Comment by nehpetsde** [comment]
(by stephen-tiedemann)
This is indeed to be considered as a bug, even if it just affects the documentation. Thanks a lot for reporting it.

**Comment by nehpetsde** [comment]
(by stephen-tiedemann)
Fix committed with trunk revision 198. 
The function assignment to thread target parameter was instead calling the function (and assigning the return value). The thread was thus not started until the LLCP link terminated.

**Comment by nehpetsde** [comment]
(by alexykot)
Hi Stephen


Yes, it works now, thank you. I've added NFC integration into my device, thanks for your work.

Have one question though. Here is a simple test stript I wrote before main application:

    import nfc, nfc.snep, threading


    nfc_thread = None
    def init_nfc(run_event, bitcoin_uri):
        clf = nfc.ContactlessFrontend('usb')
        connected = lambda llc: threading.Thread(target=llc.run).start()
        llc = clf.connect(llcp={'on-connect': connected})

        link = nfc.ndef.UriRecord(bitcoin_uri)
        snep = nfc.snep.SnepClient(llc)
        snep.put(nfc.ndef.Message(link))

        clf.close()
        if not run_event.is_set():
            return

    try:
        while True:
            if nfc_thread is None or not nfc_thread.isAlive():
                run_event = threading.Event()
                run_event.set()
                nfc_thread = threading.Thread(target=init_nfc, args=(run_event,'bitcoin:1G2bcoCKj8s9GYheqQgU5CHSLCtGjyP9Vz? amount=0.12&label=test%20merchant&message=test%20payment'))
                nfc_thread.start()

            print '!'
            time.sleep(1)
    except KeyboardInterrupt:
        run_event.clear()
        nfc_thread.join()

Problem with it is that I cannot stop the script once it executes clf.connect(). I mean - nfcpy works fine in a thread, but once it's initialized and is waiting for connection - I dont know how to stop it correctly if I need to stop the script. It catches KeyboardInterrupt exceptions and I have to either kill script or use NFC so the nfcpy execution will go though it fully and stop by itself. 

It's not a big issue in the test script, but will be a problem in my real application, as once NFC is invoked - it will be hanging on exit until killed on NFC used (although its not needed at that point). 

So how can I stop nfcpy and exit script gracefully without killing nfcpy thread?


Regards,
Alex.

**Comment by nehpetsde** [comment]
(by alexykot)
Updates on the question above:
 
a) I do need to stop nfcpy from inside my main code during normal operation cycle. So I need to sort out problem this somehow.
b) I didn't yet try to kill thread, but as far as I found it's not a pythonic way, so there is no direct support for that. Although there are ways like described here http://stackoverflow.com/questions/323972/is-there-any-way-to-kill-a-thread-in-python I'll leave it as a last resort.
c) inside nfcpy, in the clf.py file, lines 387-396, there is a `while True:` loop that actualy blocks execution. It's wrapped in KeyboardInterrupt exception, but since this whole code is executed in a daemon background thread - KeyboardInterrupt raised in the main execution doesn't reach here, and this thread runs until killed by OS or connected successfully. This is the problematic place as far as I understand. 

So in here I need some way to send a message from my main execution thread into the nfcpy thread to stop operation explicitly, not waiting for connection or exception. I can try to apply a crude patch here myself, or ideally would wait for proper patch from you.


Regards,
Alex.

**Comment by nehpetsde** [comment]
(by stephen-tiedemann)
The snippets in "Getting started" are made to fit commands on a single line, so to be copied easily into the interactive interpreter and allow going step-by-step without the need for larger function definitions. For a script I recommend to run the program logic in a new thread and have the in-built connect() loop do the housekeeping. That is something like this: http://pastebin.com/xfthYY5F where you can then press Ctrl-C to terminate earlier.

**Comment by nehpetsde** [comment]
(by alexykot)
Hmm, I see what you mean. Not sure it's the right approach though, it seems strange to me to hold all application logic in a child process and keep only NFC loop in the main execution thread. This will complicate my application development and debugging. And also in fact I don't need NFC to run all the time, NFC is an important part of my device, but it's not used literally every second device is on.

So instead I did this:
in nfcpy clf.py file:
    def connect(self, run_event, **options):
         ....
         try:
            while True:
                ....
                if not run_event.isSet():
                    return False
        except KeyboardInterrupt as error:
            ....
 

and in my application:
    def init_nfcpy(run_event, uri_to_send):
        clf = nfc.ContactlessFrontend('usb')
        connected = lambda llc: threading.Thread(target=llc.run).start()
        llc = clf.connect(run_event=run_event, llcp={'on-connect': connected})

        link_obj = nfc.ndef.UriRecord(uri_to_send)
        snep = nfc.snep.SnepClient(llc)
        snep.put(nfc.ndef.Message(link_obj))

        clf.close()

    run_event = threading.Event()
    run_event.set()
    nfc_thread = threading.Thread(target=init_nfcpy, args=(run_event, bitcoin_uri,))
    nfc_thread.start()


So basically I spawn a separate thread with nfc code, and when needed kill it on demand with `run_event.clear()` call from the main application. Not sure if my approach is good in terms of nfcpy, but it seems way more logical in terms of my development.

What do you think about this approach?

**Comment by nehpetsde** [comment]
(by stephen-tiedemann)
I see your point. There was indeed no good way to tell connect() to return earlier. I've now added a callback function parameter 'terminate' that allows application code to terminate connect() by returning True from that function. The callback function can then negotiate the return value with the application logic as needed. I've put a simplified example up at http://pastebin.com/k7L9sC1f, it uses a console interaction as the "main" logic and runs nfcpy only when needed.

Thanks for highlighting this issue.

**Comment by nehpetsde** [comment]
(by alexykot)
Great, now it works perfectly, thank you!
